pkg/api/write.go

```diff
@@ -123,31 +190,117 @@ func validateWriteHeaders(w http.ResponseWriter, r *http.Request) bool {
+			return false
+		}
+
+		if !strings.HasPrefix(remoteWriteVersion, "0.1.") {
+			buildWriteError(w, fmt.Sprintf("unexpected Remote-Write-Version %s, expected 0.1.X", remoteWriteVersion))
+			return false
+		}
+	case "application/json":
+		// Don't need any other header checks for JSON content type.
+	default:
+		buildWriteError(w, "unsupported data format (not protobuf or json)")
 		return false
 	}
 
-	if r.Header.Get("Content-Type") != "application/x-protobuf" {
-		buildWriteError(w, "non-protobuf data")
-		return false
+	return true
+}
+
+func buildWriteError(w http.ResponseWriter, err string) {
+	log.Error("msg", "Write header validation error", "err", err)
+	http.Error(w, err, http.StatusBadRequest)
+}
+
+type importPayload struct {
+	Labels  map[string]string `json:"labels"`
+	Samples []importSample    `json:"samples"`
+}
+
+type importSample struct {
+	Timestamp int64
+	Value     float64
+}
+
+func (i *importSample) UnmarshalJSON(data []byte) error {
+	var (
+		s  string
+		v  []interface{}
+		ok bool
+	)
+	if err := json.Unmarshal(data, &v); err != nil {
+		return err
+	}
+
+	if len(v) != 2 {
+		return fmt.Errorf("invalid sample, should contain timestamp and value: %v", v)
+	}
+
+	// Verify that timestamp doesn't contain fractions so we don't silently
+	// drop precision.
+	s = string(data)
+	idx := strings.IndexRune(s, '.')
+	if idx != -1 && idx < strings.IndexRune(s, ',') {
+		return fmt.Errorf("timestamp cannot contain decimal parts in sample: %s", data)
 	}
 
-	remoteWriteVersion := r.Header.Get("X-Prometheus-Remote-Write-Version")
-	if remoteWriteVersion == "" {
-		buildWriteError(w, "Missing X-Prometheus-Remote-Write-Version header")
+	floatTimestamp, ok := v[0].(float64)
+	if !ok {
+		return fmt.Errorf("cannot convert timestamp to int64 type for sample: %s", data)
+	}
+	i.Timestamp = int64(floatTimestamp)
+
+	i.Value, ok = v[1].(float64)
+	if !ok {
+		return fmt.Errorf("cannot convert value to float64 type for sample: %v", v)
+	}
+
+	return nil
+}
+
+func appendImportPayload(i importPayload, req *prompb.WriteRequest) *prompb.WriteRequest {
+	if len(i.Samples) == 0 || len(i.Labels) == 0 {
+		return req
+	}
+
+	ts := prompb.TimeSeries{}
+	ts.Labels = make([]prompb.Label, 0, len(i.Labels))
+	for name, value := range i.Labels {
+		ts.Labels = append(ts.Labels, prompb.Label{
+			Name:  name,
+			Value: value,
+		})
+	}
+
+	for _, sample := range i.Samples {
+		ts.Samples = append(ts.Samples, prompb.Sample{
+			Timestamp: sample.Timestamp,
+			Value:     sample.Value,
+		})
+	}
+
+	req.Timeseries = append(req.Timeseries, ts)
+	return req
+}
+
+func detectSnappyStreamFormat(input []byte) bool {
+	if len(input) < 10 {
 		return false
 	}
 
-	if !strings.HasPrefix(remoteWriteVersion, "0.1.") {
-		buildWriteError(w, fmt.Sprintf("unexpected Remote-Write-Version %s, expected 0.1.X", remoteWriteVersion))
+	if string(input[0:10]) != "\xff\x06\x00\x00\x73\x4e\x61\x50\x70\x59" {
```

Comment:
we should add a link to the framing format description https://github.com/google/snappy/blob/01a566f825e6083318bda85c862c859e198bd98a/framing_format.txt#L68-L81, and describe why we think filtering based on this will work.
(I'm not entirely sure it _does_ work; `0xff, 0x06` is a completely valid varint representing 895 and it may be possible to construct a valid snappy-encoded buffer starting with this string. That said, I don't think it's likely, and we can safely ignore it)

Reply:
`0xff` is the byte that signals snappy stream format, any chunk that starts with that must mean stream encoding. I doubt you can have a valid snappy buffer starting with that byte without it being a stream. I've checked the whole "magic chunk" which needs to be at the start of stream. If what you are saying is possible, it would be an issue with snappy itself IMO.

Will add a comment there to clear up what we are checking for.